### PR TITLE
stochastic extension w/ scenarios

### PR DESCRIPTION
**Pull request by oyurdakul**

### CODE REVIEW

File: src/solution/solution.jl [hunk 1]
```diff
@@ -53,66 +59,70 @@ function solution(model::JuMP.Model)::OrderedDict
...
-    sol["Down-flexiramp shortfall (MW)"] = OrderedDict(
-        r.name => [
-            value(model[:upflexiramp_shortfall][r.name, t]) for
-            t in 1:instance.time
-        ] for r in instance.reserves if r.type == "flexiramp"
-    )
+            ] for r in sc.reserves if r.type == "down-frp"
+        )
+    end
+    length(keys(sol)) > 1 ? nothing : sol = Dict(sol_key => sol_val for scen_key in keys(sol) for (sol_key, sol_val) in sol[scen_key])
```
Comment (iSoron) [inline src/solution/solution.jl:126]: This is a bit cryptic. The following would be more clear:
```julia
if length(instance.scenarios) == 1)
    return first(values(sol))
else
    return sol
end

File: src/validation/repair.jl [hunk 2]
```diff
@@ -3,19 +3,19 @@
...
-Verifies that the given unit commitment instance is valid and automatically
+Verifies that the given unit commitment scenario is valid and automatically
 fixes some validation errors if possible, issuing a warning for each error
 found. If a validation error cannot be automatically fixed, issues an
 exception.
 
 Returns the number of validation errors found.
 """
-function repair!(instance::UnitCommitmentInstance)::Int
+function repair!(sc::UnitCommitmentScenario)::Int
```
Comment (iSoron) [inline src/validation/repair.jl:15]: As discussed, it's fine to change this one to `repair!(::UnitCommitmentScenario)` but I think we should still have another one with signature `repair!(::UnitCommitmentInstance)`, which just calls this one.

**Review by iSoron — CHANGES_REQUESTED** [review]
Thanks for the PR, Ogün. Please see comments below.

**Comment by oyurdakul** [comment]
Requisite modifications to accommodate for the stochastic extension
**1. Modify validate**
- validate units [ ]
- validate reserves [ ]
**2. Modify fix! [ ]**
**3. Tests**
- usage test [ ]
- egret test [ ]
- read test [ ]
- migrate test [ ]
- formulations test [ ]
- filter test [ ]
- find test [ ]
- sensitivity test [ ]
- repair test [ ]
**4. Modify unit.jl to allow for fast-start units
**5. Eliminate redundant ISF and LODF matrices for scenarios with identical network parameters**
**6. Verify that the feasibility regions of first-stage decisions are identical across scenarios**

File: src/instance/read.jl [hunk 3]
```diff
@@ -54,18 +74,33 @@ Read instance from a file. The file may be gzipped.
 instance = UnitCommitment.read("/path/to/input.json.gz")
 ```
 """
-function read(path::AbstractString)::UnitCommitmentInstance
+function read_scenario(path::AbstractString; total_number_of_scenarios = 1, scenario_index = 1)::UnitCommitmentScenario
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/instance/read.jl]: Done

File: src/instance/read.jl [hunk 4]
```diff
@@ -54,18 +74,33 @@ Read instance from a file. The file may be gzipped.
...
 end
 
-function _read(file::IO)::UnitCommitmentInstance
+function _read(file::IO, total_number_of_scenarios::Int, scenario_index::Int)::UnitCommitmentScenario
     return _from_json(
         JSON.parse(file, dicttype = () -> DefaultOrderedDict(nothing)),
-    )
+        total_number_of_scenarios, 
+        scenario_index
+        )
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/instance/read.jl]: Done

File: src/instance/read.jl [hunk 5]
```diff
@@ -101,6 +136,17 @@ function _from_json(json; repair = true)
...
+    probability = json["Parameters"]["Scenario probability"]
+    if probability === nothing 
+        probability = (1 / total_number_of_scenarios)
+    end
+    scenario_name = json["Parameters"]["Scenario name"]
+    if scenario_name === nothing
+        scenario_name = "s$(scenario_index)"
+    end
+   
+    ######
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/instance/read.jl]: Done

File: src/model/formulations/base/unit.jl [hunk 6]
```diff
@@ -11,77 +11,135 @@ function _add_unit!(model::JuMP.Model, g::Unit, formulation::Formulation)
...
+            end
+        elseif r.type == "down-frp"
+            for t in 1:model[:instance].time
+                dwflexiramp[sc.name, r.name, g.name, t] = @variable(model) # down-flexiramp, dr_{it} in Wang & Hobbs (2016)
+                if (sc.name, r.name, t) ∉ keys(dwflexiramp_shortfall)
+                    dwflexiramp_shortfall[sc.name, r.name, t] =
+                        @variable(model, lower_bound = 0)
+                    if r.shortfall_penalty < 0
+                        set_upper_bound(dwflexiramp_shortfall[sc.name, r.name, t], 0.0)
+                    end
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/model/formulations/base/unit.jl]: Done

File: src/solution/solution.jl [hunk 7]
```diff
@@ -16,34 +16,40 @@ solution = UnitCommitment.solution(model)
...
             for b in collection
         )
     end
-    function production_cost(g)
+    function timeseries_second_stage(vars, collection, sc)
+        return OrderedDict(
+            b.name => [round(value(vars[sc.name, b.name, t]), digits = 5) for t in 1:T]
+            for b in collection
+        )
+    end
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/solution/solution.jl]: Done

File: src/solution/solution.jl (hunk 1 above)
Comment (oyurdakul) [reply src/solution/solution.jl]: Done

File: src/instance/read.jl [hunk 8]
```diff
@@ -43,6 +44,25 @@ function read_benchmark(
     return UnitCommitment.read(filename)
 end
 
+function read_scenarios(path::AbstractString)::UnitCommitmentInstance
+    scenario_paths = glob("*.json", path)
+    scenarios = Vector{UnitCommitmentScenario}()
+    total_number_of_scenarios = length(scenario_paths)
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/instance/read.jl]: Done

File: src/instance/read.jl [hunk 9]
```diff
@@ -320,14 +368,16 @@ function _from_json(json; repair = true)
         price_sensitive_loads = loads,
         reserves = reserves,
         reserves_by_name = name_to_reserve,
-        shortfall_penalty = shortfall_penalty,
-        flexiramp_shortfall_penalty = flexiramp_shortfall_penalty,
+        # shortfall_penalty = shortfall_penalty,
+        # flexiramp_shortfall_penalty = flexiramp_shortfall_penalty,
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/instance/read.jl]: Done

File: src/instance/structs.jl [hunk 10]
```diff
@@ -85,23 +87,34 @@ Base.@kwdef mutable struct UnitCommitmentInstance
     price_sensitive_loads::Vector{PriceSensitiveLoad}
     reserves::Vector{Reserve}
     reserves_by_name::Dict{AbstractString,Reserve}
-    shortfall_penalty::Vector{Float64}
-    flexiramp_shortfall_penalty::Vector{Float64}
-    time::Int
+    # shortfall_penalty::Vector{Float64}
+    # flexiramp_shortfall_penalty::Vector{Float64}
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/instance/structs.jl]: Done

File: src/model/build.jl [hunk 11]
```diff
@@ -77,20 +77,28 @@ function build_model(;
...
+            for b in scenario.buses
+                _add_bus!(model, b, scenario)
+            end
+            for ps in scenario.price_sensitive_loads
+                _add_price_sensitive_load!(model, ps, scenario)
+            end
+            for g in scenario.units
+                _add_unit_second_stage!(model, g, formulation, scenario)
+            end
+            _add_system_wide_eqs!(model, scenario)
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/model/build.jl]: Done

File: src/model/formulations/Gar1962/prod.jl [hunk 12]
```diff
@@ -22,16 +23,20 @@ function _add_production_limit_eqs!(
...
+    reserve = _total_reserves(model, g, sc)
     gn = g.name
     for t in 1:model[:instance].time
         # Objective function terms for production costs
         # Part of (69) of Kneuven et al. (2020) as C^R_g * u_g(t) term
-        add_to_expression!(model[:obj], is_on[gn, t], g.min_power_cost[t])
+
+        ### Moving this term to another function
+        # add_to_expression!(model[:obj], is_on[gn, t], g.min_power_cost[t])
+        ###
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/model/formulations/Gar1962/prod.jl]: Done

File: src/model/formulations/WanHob2016/ramp.jl [hunk 13]
```diff
@@ -29,7 +30,7 @@ function _add_ramp_eqs!(
         error("Each generator may only provide one flexiramp reserve")
     end
     for r in g.reserves
-        if r.type !== "flexiramp"
+        if r.type !== "up-frp" && r.type !== "down-frp"
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/model/formulations/WanHob2016/ramp.jl]: Done

File: src/model/formulations/base/line.jl [hunk 14]
```diff
@@ -22,27 +23,28 @@ end
 function _setup_transmission(
     model::JuMP.Model,
     formulation::ShiftFactorsFormulation,
+    scenario::UnitCommitmentScenario
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/model/formulations/base/line.jl]: Done

File: src/model/formulations/base/unit.jl [hunk 15]
```diff
@@ -11,77 +11,135 @@ function _add_unit!(model::JuMP.Model, g::Unit, formulation::Formulation)
...
+#         formulation.prod_vars,
+#         formulation.ramping,
+#         formulation.status_vars,
+#     )
+#     _add_startup_cost_eqs!(model, g, formulation.startup_costs)
+#     _add_startup_shutdown_limit_eqs!(model, g)
+#     _add_status_eqs!(model, g, formulation.status_vars)
+#     return
+# end
+
```
[the remark this answers is not in the document]
Comment (oyurdakul) [reply src/model/formulations/base/unit.jl]: Done

**Pull request merged**